Any match failures trigger panic

A match failure on a ref of type any triggers a panic:

```
x = ["foo", 100]; x[_] = null
```

Panic:

```
panic: unreachable [recovered]
	panic: unreachable

goroutine 65 [running]:
testing.tRunner.func1(0xc4204648f0)
	/usr/local/Cellar/go/1.8.3/libexec/src/testing/testing.go:622 +0x29d
panic(0x1235c40, 0xc42044b530)
	/usr/local/Cellar/go/1.8.3/libexec/src/runtime/panic.go:489 +0x2cf
github.com/open-policy-agent/opa/ast.unifies(0x1390f40, 0xc420467b60, 0x1390fc0, 0x13d2378, 0xc4204678c0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:682 +0x195
github.com/open-policy-agent/opa/ast.unify2(0xc42044b450, 0xc4204678a0, 0x1390f40, 0xc420467b60, 0xc4204678c0, 0x1390fc0, 0x13d2378, 0x2)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:285 +0x926
github.com/open-policy-agent/opa/ast.(*typeChecker).checkExprEq(0xc4204679a0, 0xc42044b450, 0xc4203dc4b0, 0xc420084a90)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:263 +0x14c
github.com/open-policy-agent/opa/ast.(*typeChecker).(github.com/open-policy-agent/opa/ast.checkExprEq)-fm(0xc42044b450, 0xc4203dc4b0, 0x1280fd1)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:33 +0x3e
github.com/open-policy-agent/opa/ast.(*typeChecker).checkExpr(0xc4204679a0, 0xc42044b450, 0xc4203dc4b0, 0xc4203dc4b0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:205 +0xc6
github.com/open-policy-agent/opa/ast.(*typeChecker).CheckBody.func1(0xc4203dc4b0, 0x1170901)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:66 +0x220
github.com/open-policy-agent/opa/ast.WalkExprs.func1(0x1277640, 0xc4203dc4b0, 0x1170901)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/visit.go:127 +0x42
github.com/open-policy-agent/opa/ast.(*GenericVisitor).Visit(0xc42008c120, 0x1277640, 0xc4203dc4b0, 0x0, 0x0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/visit.go:209 +0x3d
github.com/open-policy-agent/opa/ast.Walk(0x138f940, 0xc42008c120, 0x1277640, 0xc4203dc4b0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/visit.go:18 +0x66
github.com/open-policy-agent/opa/ast.Walk(0x138f940, 0xc42008c120, 0x1268f40, 0xc420467a00)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/visit.go:56 +0xacd
github.com/open-policy-agent/opa/ast.WalkExprs(0x1268f40, 0xc420467a00, 0xc4204679e0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/visit.go:131 +0xb5
github.com/open-policy-agent/opa/ast.(*typeChecker).CheckBody(0xc4204679a0, 0xc42044b450, 0xc42044b420, 0x2, 0x2, 0xc420083400, 0xc420023900, 0x0, 0x0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check.go:78 +0x2a5
github.com/open-policy-agent/opa/ast.TestCheckMatchErrors.func1(0xc4204648f0)
	/Users/torin/go/src/github.com/open-policy-agent/opa/ast/check_test.go:556 +0xa0
testing.tRunner(0xc4204648f0, 0xc42007b0b0)
	/usr/local/Cellar/go/1.8.3/libexec/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/local/Cellar/go/1.8.3/libexec/src/testing/testing.go:697 +0x2ca
exit status 2
```

--- a/ast/check.go
+++ b/ast/check.go
@@ -630,18 +630,22 @@ func unifies(a, b types.Type) bool {
 		return false
 	}
 
-	var unified bool
-
-	if anyA, ok := a.(types.Any); ok {
-		unified = unifiesAny(anyA, b)
+	anyA, ok1 := a.(types.Any)
+	if ok1 {
+		if unifiesAny(anyA, b) {
+			return true
+		}
 	}
 
-	if anyB, ok := b.(types.Any); ok {
-		unified = unified || unifiesAny(anyB, a)
+	anyB, ok2 := b.(types.Any)
+	if ok2 {
+		if unifiesAny(anyB, a) {
+			return true
+		}
 	}
 
-	if unified {
-		return true
+	if ok1 || ok2 {
+		return false
 	}
 
 	switch a := a.(type) {

--- a/ast/check_test.go
+++ b/ast/check_test.go
@@ -547,6 +547,7 @@ func TestCheckMatchErrors(t *testing.T) {
 		{"object-nested", `{ {"a": 1, "b": "2"} = {"a": 1, "b": 2} }`},
 		{"object-nested-2", `{ {"a": 1} = {"a": 1, "b": "2"} }`},
 		{"set", "{{1,2,3} = null}"},
+		{"any", `{x = ["str", 1]; x[_] = null}`},
 	}
 	for _, tc := range tests {
 		test.Subtest(t, tc.note, func(t *testing.T) {

